File: lua/luhn.lua
return {
    valid = function(s)
        if s:match("[^0-9 ]") then
            return false
        end

        if #s < 2 then
            return false
        end

        local flag = #s % 2 == 0
        local sum = 0
        for c in s:gmatch("%S") do
            local n = c - "0"
            if flag then
                n = n * 2
                if n > 9 then
                    n = n - 9
                end
            end
            sum = sum + n
            flag = not flag
        end

        return sum % 10 == 0
    end,
}

File: ocaml/luhn.ml
open Base

let valid num =
  let lst =
    num |> String.to_list |> List.filter ~f:(Fn.compose not (Char.equal ' '))
  in
  if List.length lst < 2 then false
  else
    match List.find ~f:(Fn.compose not Char.is_digit) lst with
    | Some _ -> false
    | None ->
        List.rev lst
        |> List.rev_mapi ~f:(fun i n ->
               let n = Char.get_digit_exn n in
               if i % 2 = 1 then
                 let n = n * 2 in
                 if n > 9 then n - 9 else n
               else n)
        |> List.fold_left ~init:0 ~f:( + )
        |> Fn.flip ( % ) 10
        |> ( = ) 0
